### PR TITLE
bcc: devendor libbpf

### PR DESCRIPTION
**Pull request by martinetd**
###### Motivation for this change

Follow-up cleanup after discussion in #116588

###### /!\ Warning /!\

While this does build on linuxPackage_latest, this will be much pickier about building "against" older linux kernels than it used to be.
The rationale is that libbpf when vendored also in turn vendored linux headers appropriate for the libbpf version considered, but these are no longer available so we depend on every latest libbpf feature being present in the associated linux headers.

Linux is careful about ABI breakage so in practice using vendored libbpf headers was a good idea: it doesn't harm to have newer includes that define strictly more enum values and the code has runtime detection (well, trying to use newer features will just fail with ENOTSUP) -- it might actually make sense to have libbpf install its vendored linux uapi headers as well as its regular headers and have bcc/other bpf tools use these to avoid dependency hell...

TL;DR: with this patch, bcc will now only compile on linux kernels newer than bcc release date unless bcc is patched conditionally (e.g. to build with a 5.10 kernel removing a single line is enough: https://gitlab.alpinelinux.org/alpine/aports/-/raw/a55371125e8fd13800e8a394c80a4a4f930af64f/community/bcc/fix-newer-kernel-header-missing-enum.patch )

With that said it might be better to keep the current vendoring...

Cc @mic92 @ragnard @thoughtpolice (maintainers)

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by SuperSandro2000** [comment]
This is a semi-automatic executed nixpkgs-review with [nixpkgs-review-checks extension](https://github.com/SuperSandro2000/nixpkgs-review-checks). It is checked by a human on a best effort basis and does not build all packages (e.g. lumo, tensorflow or pytorch).
If you have any questions or problems please reach out to SuperSandro2000 on IRC.

Result of `nixpkgs-review pr 118121` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
<summary>
10 packages marked as broken and skipped:
</summary>
<ul>
<li>linuxPackages-libre.bcc</li>
<li>linuxPackages-libre.bpftrace</li>
<li>linuxPackages-libre.oci-seccomp-bpf-hook</li>
<li>linuxPackages_hardkernel_4_14.bcc</li>
<li>linuxPackages_hardkernel_4_14.bpftrace</li>
<li>linuxPackages_hardkernel_latest.bcc</li>
<li>linuxPackages_hardkernel_latest.bpftrace</li>
<li>linuxPackages_latest-libre.bcc</li>
<li>linuxPackages_latest-libre.bpftrace</li>
<li>linuxPackages_latest-libre.oci-seccomp-bpf-hook</li>
</ul>
</details>
<details>
<summary>
44 packages built:
</summary>
<ul>
<li>linuxPackages.bcc (linuxPackages_5_10.bcc)</li>
<li>linuxPackages.bpftrace (linuxPackages_5_10.bpftrace)</li>
<li>linuxPackages.oci-seccomp-bpf-hook (linuxPackages_5_10.oci-seccomp-bpf-hook)</li>
<li>linuxPackages_4_14.bcc</li>
<li>linuxPackages_4_14.bpftrace</li>
<li>linuxPackages_4_19.bcc</li>
<li>linuxPackages_4_19.bpftrace</li>
<li>linuxPackages_4_4.bcc</li>
<li>linuxPackages_4_4.bpftrace</li>
<li>linuxPackages_4_9.bcc</li>
<li>linuxPackages_4_9.bpftrace</li>
<li>linuxPackages_5_11.bcc (linuxPackages_latest.bcc)</li>
<li>linuxPackages_5_11.bpftrace (linuxPackages_latest.bpftrace)</li>
<li>linuxPackages_5_11.oci-seccomp-bpf-hook (linuxPackages_latest.oci-seccomp-bpf-hook)</li>
<li>linuxPackages_5_4.bcc</li>
<li>linuxPackages_5_4.bpftrace</li>
<li>linuxPackages_5_4.oci-seccomp-bpf-hook</li>
<li>linuxPackages_hardened.bcc</li>
<li>linuxPackages_hardened.bpftrace</li>
<li>linuxPackages_hardened.oci-seccomp-bpf-hook</li>
<li>linuxPackages_latest_hardened.bcc</li>
<li>linuxPackages_latest_hardened.bpftrace</li>
<li>linuxPackages_latest_hardened.oci-seccomp-bpf-hook</li>
<li>linuxPackages_latest_xen_dom0.bcc</li>
<li>linuxPackages_latest_xen_dom0.bpftrace</li>
<li>linuxPackages_latest_xen_dom0.oci-seccomp-bpf-hook</li>
<li>linuxPackages_latest_xen_dom0_hardened.bcc</li>
<li>linuxPackages_latest_xen_dom0_hardened.bpftrace</li>
<li>linuxPackages_latest_xen_dom0_hardened.oci-seccomp-bpf-hook</li>
<li>linuxPackages_lqx.bcc</li>
<li>linuxPackages_lqx.bpftrace</li>
<li>linuxPackages_lqx.oci-seccomp-bpf-hook</li>
<li>linuxPackages_testing_bcachefs.bcc</li>
<li>linuxPackages_testing_bcachefs.bpftrace</li>
<li>linuxPackages_testing_bcachefs.oci-seccomp-bpf-hook</li>
<li>linuxPackages_xen_dom0.bcc</li>
<li>linuxPackages_xen_dom0.bpftrace</li>
<li>linuxPackages_xen_dom0.oci-seccomp-bpf-hook</li>
<li>linuxPackages_xen_dom0_hardened.bcc</li>
<li>linuxPackages_xen_dom0_hardened.bpftrace</li>
<li>linuxPackages_xen_dom0_hardened.oci-seccomp-bpf-hook</li>
<li>linuxPackages_zen.bcc</li>
<li>linuxPackages_zen.bpftrace</li>
<li>linuxPackages_zen.oci-seccomp-bpf-hook</li>
</ul>
</details>

**Comment by SuperSandro2000** [comment]
@martinetd seems to build fine for all kernels.

**Review by SuperSandro2000 — APPROVED** [review]

**Comment by martinetd** [comment]
@SuperSandro2000 
> seems to build fine for all kernels.

That's ... surprising? it's using e.g. `BPF_MAP_TYPE_TASK_STORAGE` that should come with 5.11 kernel headers?!

. . .
. . .
ok so it looks like glibc also redefines the same types, and bcc ends up using these instead of linux's headers, so this works fine.
I'm appalled at the duplication there but it works in our favor, so I'm happy with this PR now. Thanks for checking, I was so convinced it wouldn't work that I didn't take the time to try..

**Comment by thoughtpolice** [comment]
It's possible I've misunderstood something but I'm not inclined to accept this patch I'm afraid. It simply removes support for many kernels and users, and has little upside. When it still builds where it shouldn't, it's discovered this is only the case due to some spooky action-at-a-distance interaction between the include priorities of glibc, the linux kernel, and libbpf, and all of them duplicating things, apparently? This is all just too fragile to rely on I'm afraid, unless we have a solid grasp of *what* is supposed to be happening.

I understand the desire to remove vendored dependencies in general, but the solution here seems to be unequivocally worse than the problem itself.

**Review by thoughtpolice — CHANGES_REQUESTED** [review]
(See prior comment.)

**Comment by martinetd** [comment]
> It's possible I've misunderstood something but I'm not inclined to accept this patch. It simply removes support for many kernels and users, and has little upside. When it still builds where it shouldn't, it's discovered this is only the case due to some spooky action-at-a-distance interaction between the include priorities of glibc, the linux kernel, and libbpf, and all of them duplicating things, apparently? This is all just too fragile to rely on I'm afraid, unless we have a solid grasp of what is supposed to be happening.

If it builds for all kernels it doesn't remove support for anything, and it doesn't look fragile -- 
in this case the only confusion was in my head when I did the original post.
I thought bcc was kernel dependent because kernel headers ultimately come from the kernel, but in nixos cases kernel headers are packaged in glibc's output directory (hence me saying it comes with glibc, it's not actually duplicated in glibc sources) and the only reason bcc is a kernel package is they need to know the module path... Which is actually a bad thing because that means you can't use a bcc for a different kernel if you upgrade your system without rebooting :|

I think bcc should be moved out of kernelPackages and just use `/run/booted-system/kernel-modules/lib/modules/` for cmake's `-DBCC_KERNEL_MODULES_DIR` like systemd does... Anyway that's out of scope of this PR and I won't put my hands into that.


So basically even bcc in linuxPackages_4_14 use linux 5.11 kernel headers, and things just work™ because the kernel doesn't break its API:
```
$ nix-shell -A linuxPackages_5_4.bcc
...
$ rm -rf /tmp/bcc && mkdir /tmp/bcc && cd /tmp/bcc && unpackPhase && cd source
$ cmakeConfigurePhase && cd build && make VERBOSE=1
...
$ cd /tmp/bcc/source/build/introspection && /nix/store/gdg97kdw8k7nqg704g4gm3cn8280bmgr-gcc-wrapper-10.2.0/bin/gcc  -I/nix/store/cval94fchij6pmjszffcyciq26xmd638-llvm-7.1.0/include/../tools/clang/include -I/tmp/bcc/source/src/cc -I/tmp/bcc/source/src/cc/api -I/tmp/bcc/source/src/cc/libbpf/include/uapi -Wall -O3 -DNDEBUG -E /tmp/bcc/source/introspection/bps.c | grep bpf.h
...
/nix/store/hnpwygsgcwxq15za8kgc3qvnh9rpwzp0-glibc-2.32-39-dev/include/linux/bpf.h
...
$ nix-store --query -R /nix/store/hnpwygsgcwxq15za8kgc3qvnh9rpwzp0-glibc-2.32-39-dev | grep linux-header
/nix/store/vrnaqvg51zwh8y2b4y6cbj0didk467gp-linux-headers-5.11
```

So basically the only change wrt headers this bring is that we need linux-headers to be roughly at least as up to date as the version vendored in libbpf source tree; given how up to date we are I don't see that as much of a problem for master but it can be more annoying for stable branches perhaps if you ever update bcc/bpf there?

The second "challenge" this change brings is that bcc now no longer picks the "correct" version of libbpf, so if the API changes builds can also stop working. I don't expect libbpf upgrades to break bcc too often but bcc updates will most likely often need libbpf upgrade as new functions are added for new types  (e.g. kernel added support for btf floats so libbpf gained `btf__add_float` and if bcc starts using it it'll need a new libbpf upgrade) -- but that's how all dependencies work, and other distros have the same requirements (and e.g. debian is a first class citizen for bcc), they will probably make sure this is workable.


Anyway shrug I only did that after being told off for just noting pahole vendors libbpf where pahole doesn't yet have a built-in way of building with distro's version, and pointed out there are other software like bcc that have a proper way of doing it and still vendor it... If y'all want to fight over it have fun.

**Comment by Mic92** [comment]
The points of @martinetd sounds reasonable. We cannot use `/run/booted-system/kernel-modules/` however in the current form however since it does not contain kernel headers right now but only kernel modules. There are two solutions to this problem so
that we can apply in our nixos bcc module:

1. Load the `kheaders` by default
2. Install kernel headers into system closure so they are also available at `/run/booted-system/kernel-modules/`

**Comment by martinetd** [comment]
> We cannot use /run/booted-system/kernel-modules/ however in the current form however since it does not contain kernel headers right now but only kernel modules.

I made the same mistake at the start of this thread, but kernel headers do NOT come from kernelPackages : they are packaged in the glibc package, and all kernel builds regardless of version use whatever is in the current glibc (so right now even 5.4 or 4.14 kernels bcc will build with kernel 5.11 kernel headers thanks to this packaging)

This also solves the problem of run-time bcc usage that also needs kernel headers, they can use whatever is in glibc's

**Comment by Mic92** [comment]
Works for me!

**Comment by r-burns** [comment]
So libbpf added BTF_KIND_FLOAT support (https://github.com/libbpf/libbpf/commit/471e7c241d30df733e51e75be39040c8da175e42) which requires Linux 5.14, but our glibc headers are pre-5.14 so bcc is failing in staging-next (https://github.com/libbpf/libbpf/commit/471e7c241d30df733e51e75be39040c8da175e42). IMO it would be simplest to revert this so that libbpf is vendored again.

**Comment by martinetd** [comment]
(replying in the re-vendor PR as not to duplicate the discussion)
